### ISSUE TITLE
Native globbing doesn't respect bypassing via the escape character (backtick) in unquoted arguments

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#15959]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When calling external programs on Unix-like platforms, native globbing can be bypassed on demand with single- or double-quoting an argument; e.g., `/bin/echo '*'`

The same should work with selectively escaping globbing metacharacters with `` ` `` in _unquoted_ arguments, but currently doesn't.

```powershell
# On Unix
/bin/echo `*
```

### Expected behavior

```console
*
```

### Actual behavior

```console
The names of all files and directories are output, because globbing was unexpectedly still applied.
```

### Environment data

```powershell
PowerShell Core 7.2.0-preview.8 on Unix-like platforms.
```

**Comment by 237dmitry** [comment]
Why expected behavior is `*` output? In bash `echo *` outputs string with all files and directories in current location (space separated). In pwsh ```/bin/echo `*``` outputs string of unhidden files and directories

**Comment by mklement0** [comment]
Because `` `* `` is an _escaped_ `*` character, to be used _verbatim_ - just like `'*'`, which _does_ bypass native globbing.

**Comment by 237dmitry** [comment]
Escaping is not equivalent quoting. You pass the escaped character to receive it as is. So that it is not lost or it is not interpreted in some other way. So `/bin/echo` gets everything as it is. Maybe I'm wrong, but that's how I understand this.

**Comment by mklement0** [comment]
Yes, escaping is equivalent to quoting  - just two different ways of achieving the same effect; compare `` Write-Output `| ``  to `Write-Output '|'`

> You pass the escaped character to receive it as is

That's precisely the point, and passing `*` as-is means just that: the external executable `/bin/echo` should receive `*` verbatim and therefore also print it verbatim.

Note that _no (native) shell is involved_ in such a call.
 
It is _PowerShell_ that _emulates_ a feature of POSIX-compatible shells - hence the name native globbing - and the external utility being invoked sees the _results_ of that. And if I'm telling _PowerShell_ to treat a character verbatim, it should therefore not apply this emulation - irrespective of the syntax form used.

**Comment by 237dmitry** [comment]
> compare ```Write-Output `|``` to `Write-Output '|'`

```powershell
PS > /bin/echo `|
|
PS >
```
Perhaps the  `*` character is a special character for `/bin/echo`, which should еxplicitly be in quotation marks.

**Comment by mklement0** [comment]
No, `*` has no special meaning to `echo`.

This is purely about _PowerShell_ respecting the explicitly signaled intent to treat a character verbatim, which _if unquoted or unescaped_ would have special meaning (a _situational_ metacharacter that only acts a such when (a) calling _external programs_ and (b) on Unix-like platforms only).

**Comment by mklement0** [comment]
And, to be clear, programs on Unix receive arguments as an array of verbatim values - they have no information about what particular shell with what particular quoting/escaping syntax was used to translate a command line into that array of verbatim values. For instance, ``/bin/echo `|``, `/bin/echo '|'` and `/bin/echo "|"` all result in the _same_ invocation of `/bin/echo`, namely with a single-element argument array whose only element contains verbatim `|`.

**Comment by vexx32** [comment]
WG-Engine discussed this issue in today's meeting.

Our opinion is that this issue with wildcards is not exclusive to the native command globbing behaviour &mdash; although, yes, the ability to use quotes to quietly bypass the globbing behaviour is specific to this experimental feature.

Given that there is currently a workaround that will likely suffice for most use cases, which @mklement0 outlines in the initial post, we would prefer effort be spent investigating more complete fixes for the issues with wildcard escaping that currently exist, rather than trying to design exceptions to the wildcard handling specifically for this feature. If a robust solution can be found to that issue (see #7999; there is also some relevant discussion in #7399), it would become trivial to incorporate the solution into the behaviour of this feature as well.

Ideally the functionality of this feature should match the behaviour of wildcard globbing as implemented in provider cmdlets.

Given that, we don't think this specific issue makes sense to fix in isolation, and it would be likely to result in additional confusion when such a workaround would subsequently fail to work with wildcards in other areas of pwsh.

**Comment by vexx32** [comment]
Resolving this one as a duplicate of #7999 as the root cause is ultimately the wildcard handling.

**Comment by mklement0** [comment]
Just to note an important distinction, @vexx32:

* #7999 is about how _cmdlets / providers_ treat wildcard expressions that they see _after_ the parameter binder has bound argument. That is, `Get-Item *`, ``Get-Item `*`` (the bareword parsing eats the `` ` ``), `Get-Item '*'`, and `Get-Item "*"` are - and should be - equivalent, because `Get-Item` sees verbatim `*` in all cases.

* By contrast, we're talking about an _intentional anomaly when calling external programs on Unix_, in service of emulating a different shell's behavior: as in POSIX-compatible shells, what matters is whether the argument is _quoted_ / has _individual metacharacters escaped_ (e.g., _from Bash_, compare `echo *` with `echo \*` / `echo '*'` / `echo "*"`), in which case _no_ globbing must be applied. 
  * Therefore, the solution must be implemented at the level of the code that parses the raw argument and decides whether native globbing is to be performed.

**Comment by vexx32** [comment]
Yeah, I'd agree that the code path here is somewhat separate to the one originally mentioned in #7999. However, the problem is likely caused by similar handling of input -- both use WildcardPattern -- and I think we were all more or less in agreement that it would counterintuitive to implement a partial fix that only addresses this issue instead of generalising the fix to also resolve the original issue for #7999.

I think it's definitely worth mentioning that in #7999 and we can incorporate this facet into the original issue as well. 🙂
